ui/collapsible-panel.reel/collapsible-panel.js

```diff
@@ -0,0 +1,126 @@
+     * header to collapse or expand a panel by setting this property to true.
+     *
+     * @private
+     */
+    _isTransition: {
+        value: false
+    },
+
+    _isExpanded: {
+        value: null
+    },
+
+    /**
+     * CollapsiblePanel's data.
+     *
+     * @type {Object}
+     */
+    data: {
+        value: null
+    },
+
+    /**
+     * A ChoiceController that control expanded panel and multiple mode of CollapsiblePanel.
+     *
+     * @type {Object}
+     */
+    expansionController: {
+        set: function (value) {
+            if (value instanceof ChoiceController) {
+                this._expansionController = value;
+                if (this.getBinding("isExpanded")) {
+                    this.cancelBinding("isExpanded");
+                }
+                this.defineBinding("isExpanded", {
+                    "<->": "_expansionController.choice.data.has(data)"
+                });
+            } else {
+                throw new Error("expansionController must be a ChoiceController.");
+            }
+        }
+    },
+
+    /**
+     * Is the collapsible panel expanded or collapsed.
+     *
+     * @type {boolean}
+     */
+    isExpanded: {
+        set: function (value) {
+            if (value != this._isExpanded) {
+                this._isExpanded = value;
+                this.needsDraw = true;
+            }
+        },
+        get: function () {
+            return this._isExpanded;
+        }
+    },
+
+    prepareForActivationEvents: {
+        value: function () {
+            var handler = this.handleTransitionEnd.bind(this);
+            var bodyWrapper = this.element.children[1];
+            this._pressComposer.addEventListener("press", this, false);
+            // Listen to CSS transition call back.
+            bodyWrapper.addEventListener('webkitTransitionEnd', handler);
+            bodyWrapper.addEventListener('transitionend', handler);
+        }
+    },
+
+    handleTransitionEnd: {
+        value: function () {
+            this._isTransition = false;
+        }
+    },
+
+    /**
+     * Do expand or collapse when click on the header.
+     *
+     * @method
+     * @private
+     */
+    handlePress: {
+        value: function (event) {
+            // Only trigger css animation when clicking on header, this may need change base on how SelectionController
+            // handle multipleSelection and singleSelection.
+            this._isTransition = true;
+            this.isExpanded = !this.isExpanded;
+        }
+    },
+
+    enterDocument: {
+        value: function (firstTime) {
+            var header;
+            if (firstTime) {
+                header = this.element.children[0];
+                this._pressComposer = new PressComposer();
+                // Register composer here and add event in prepareForActivationEvents.
+                this.addComposerForElement(this._pressComposer, header);
+            }
+        }
+    }
+
+});
```

Comment:
@dharcourt I removed instanceof checking for ChoiceController type so we allow people to pass in their own Controller. I checked this._expansionController.hasOwnProperty('choice') && this._expansionController.choice.hasOwnProperty('data') to make sure isExpanded binding can work correctly.  I think we can't allow user to binding isExpand outside because we bind it when setting expansionController and another binding will cause duplicate binding exception. In my opinion, we may do one way or another, 1) isExpanded is binding when setting _expansionController and we allow people to do their own _expansionController like our current way. I changed isExpanded to private too, so only use expansionController. 2) we binding isExpanded outside like the old way, don't do _expansionController but directly bind isExpanded outside so user can more free when using isExpanded

PS: I checked that montage uses lots instancef in other classes so I guess it's a correct way to use if we really want to check a instance Class type.